### PR TITLE
Using Monte-Carlo Approach for the Producer-Consumer Channel Benchmark 

### PR DESCRIPTION
**Pull request by fmixing**
The development mode works under 10 minutes now on macbook pro 13" (2017): BUILD SUCCESSFUL in 8m 38s.
You can see the results on AWS m5.24xlarge in the attached pdf file.
[channel-prod-cons-monte-carlo.pdf](https://github.com/Kotlin/kotlinx.coroutines/files/4266638/channel-prod-cons-monte-carlo.pdf)

### CODE REVIEW

File: benchmarks/build.gradle [hunk 1]
```diff
@@ -43,6 +43,11 @@ task removeRedundantFiles(type: Delete) {
     delete "$buildDir/classes/kotlin/jmh/benchmarks/flow/misc/Numbers\$\$special\$\$inlined\$filter\$1\$1.class"
 }
 
+task runMonteCarloBasedChannelBenchmark(type : JavaExec) {
```
Comment (ndkoval) [inline benchmarks/build.gradle]: This is a producer-consumer channel benchmark, which uses the Monte-Carlo method

File: benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import kotlinx.coroutines.channels.Channel
+
+/**
+ * Total amount of threads in a benchmark
+ */
+val THREADS = arrayOf(1, 2, 4, 8, 16, 18, 32, 36, 64, 72, 96, 108, 128, 144)
+/**
+ * Coroutines channels type.
+ */
+val CHANNEL_TYPES = listOf(ChannelType.BUFFERED_DEFAULT, ChannelType.RENDEZVOUS, ChannelType.UNLIMITED, ChannelType.BUFFERED_1, ChannelType.BUFFERED_16, ChannelType.BUFFERED_256)
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt]: Let's use the enum from the `ProducerConsumerChannelBenchmark`

Comment (ndkoval) [reply benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt]: Moreover, let's combine them!

File: benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+ */
+const val BASELINE_WORK = 50
+/**
+ * The max multiplier for the [BASELINE_WORK]
+ */
+const val MAX_WORK_MULTIPLIER = 5.0
+/**
+ * The amount of different numbers of loop iterations threads can do on average
+ */
+const val WORK_TYPES = 3
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt]: How do you use this?

File: benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ * Approximate total number of sent/received messages
+ */
+const val APPROX_BATCH_SIZE = 750_000
+/**
+ *
+ */
+const val BENCHMARK_CONFIGURATION_STOP_THRESHOLD = 0.01
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt]: What does this `0.01` mean?

File: benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+/**
+ *
+ */
+const val ITERATIONS_BETWEEN_THRESHOLD_CHECK = 50
+/**
+ * Benchmark output file
+ */
+const val OUTPUT = "out/montecarlo.out"
+
+enum class ChannelType {
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt]: Why do you need both `CHANNEL_TYPES` and this enum?

File: benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    RENDEZVOUS {
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt]: You can pass a function to construct the corresponding channel as a constructor parameter; thus, the code becomes shorter and easier to understand

File: benchmarks/src/main/kotlin/montecarlo/MonteCarloBasedChannelBenchmark.kt [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+                doWork(work)
+            }
+            phaser.arrive()
+        }.start()
+    }
+
+    return startTime
+}
+
+fun createWorkLoad(producers: Int, consumers: Int): Pair<List<Int>, List<Int>> {
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/montecarlo/MonteCarloBasedChannelBenchmark.kt]: Do we really need to balance the system for coroutines? Please, write both pros and cons, so that we can decide.

File: benchmarks/src/main/kotlin/montecarlo/MonteCarloBasedChannelBenchmark.kt [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+import java.nio.file.Paths
+import java.util.concurrent.Phaser
+import java.util.concurrent.ThreadLocalRandom
+import kotlin.math.abs
+import kotlin.math.max
+import kotlin.math.pow
+import kotlin.math.sqrt
+import kotlin.random.Random
+
+fun main() {
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/montecarlo/MonteCarloBasedChannelBenchmark.kt]: Please, add documentation with a detailed description of how the benchmark works

File: benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt [hunk 9]
```diff
@@ -0,0 +1,83 @@
+package montecarlo
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/montecarlo/BenchmarkConfiguration.kt]: why this package?

File: benchmarks/src/main/kotlin/channel/ChannelProducerConsumerMonteCarloBenchmark.kt [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+    // start all threads
+    val channelProducerConsumerBenchmarkWorker = ChannelProducerConsumerBenchmarkWorkerMonteCarlo(
+            selectMode == BenchmarkSelectMode.WITH_SELECT, producersDispatcher, producerConsumerWorkLoad.producerWorks,
+            consumersDispatcher, producerConsumerWorkLoad.consumerWorks, channelCreator)
+
+    val startTime = System.nanoTime()
+    channelProducerConsumerBenchmarkWorker.run(producers, consumers, APPROXIMATE_BATCH_SIZE)
+    val endTime = System.nanoTime()
+
+    if (producersDispatcher is Closeable) {
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/channel/ChannelProducerConsumerMonteCarloBenchmark.kt]: `(disparcher is Closeable)?.close()

File: benchmarks/src/jmh/kotlin/benchmarks/ChannelProducerConsumerBenchmark.kt [hunk 11]
```diff
@@ -46,101 +51,40 @@ open class ChannelProducerConsumerBenchmark {
...
 private const val WORK_MAX = 100L
-private const val APPROX_BATCH_SIZE = 100000
\ No newline at end of file
+private const val APPROX_BATCH_SIZE = 100000
+
+class ChannelProducerConsumerBenchmarkWorkerJMH(withSelect: Boolean,
+                                                dispatcher: CoroutineDispatcher,
+                                                channelCreator: ChannelCreator)
+    : ChannelProducerConsumerBenchmarkWorker(withSelect, dispatcher, dispatcher, channelCreator) {
+    override fun doProducerWork(coroutineNumber: Int) = Blackhole.consumeCPU(ThreadLocalRandom.current().nextLong(WORK_MIN, WORK_MAX))
```
Comment (ndkoval) [inline benchmarks/src/jmh/kotlin/benchmarks/ChannelProducerConsumerBenchmark.kt]: use geometric distribution here

File: benchmarks/src/main/kotlin/ChannelProducerConsumerBenchmarkWorker.kt [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.selects.select
+import java.util.concurrent.ForkJoinPool
+import java.util.concurrent.Phaser
+
+abstract class ChannelProducerConsumerBenchmarkWorker(private val withSelect: Boolean,
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/ChannelProducerConsumerBenchmarkWorker.kt]: Let's rename it `ChannelProducerConsumerBenchmarkIteration` and pass `producers` and `consumers` as parameters so that it has a simple `run()` function.

File: benchmarks/src/main/kotlin/BenchmarkUtils.kt [hunk 13]
```diff
@@ -0,0 +1,14 @@
+/*
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/BenchmarkUtils.kt]: missed package

File: benchmarks/src/jmh/kotlin/benchmarks/ChannelProducerConsumerBenchmark.kt [hunk 14]
```diff
@@ -47,100 +51,44 @@ open class ChannelProducerConsumerBenchmark {
...
-//    BUFFERED_1(1),
-    BUFFERED_2(2),
-//    BUFFERED_4(4),
-    BUFFERED_32(32),
-    BUFFERED_128(128),
-    BUFFERED_UNLIMITED(Channel.UNLIMITED);
-
-    fun create(): Channel<Int> = Channel(capacity)
-}
+private const val WORK = 40
```
Comment (ndkoval) [inline benchmarks/src/jmh/kotlin/benchmarks/ChannelProducerConsumerBenchmark.kt]: Let's change this constant to `80`

File: benchmarks/src/jmh/kotlin/benchmarks/ChannelProducerConsumerBenchmark.kt [hunk 15]
```diff
@@ -47,100 +51,44 @@ open class ChannelProducerConsumerBenchmark {
...
-        channel = _1_channel.create()
     }
 
     @Benchmark
     fun spmc() {
         if (_2_coroutines != 0) return
         val producers = max(1, _4_parallelism - 1)
         val consumers = 1
-        run(producers, consumers)
+        val iterationJMH = ChannelProducerConsumerBenchmarkIterationJMH(_3_withSelect, dispatcher, _1_channel, producers,
```
Comment (ndkoval) [inline benchmarks/src/jmh/kotlin/benchmarks/ChannelProducerConsumerBenchmark.kt]: `iterationJMH` -> `iteration`

File: benchmarks/src/main/kotlin/benchmarks/channel/ChannelProducerConsumerMonteCarloBenchmark.kt [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+ */
+private val WITH_BALANCING = listOf(true, false)
+/**
+ * Should select be used in the benchmark or not
+ */
+private val WITH_SELECT = listOf(true, false)
+/**
+ * Benchmark output file
+ */
+private const val OUTPUT = "out/resultsProdCons.csv"
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/benchmarks/channel/ChannelProducerConsumerMonteCarloBenchmark.kt]: let's rename the output file to "results_channel_producer_consumer_montecarlo.csv"

File: benchmarks/src/main/kotlin/benchmarks/channel/ChannelProducerConsumerMonteCarloBenchmark.kt [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+ * workload is randomly chosen (workload may be balanced or unbalanced according to the benchmark configuration). Then
+ * the benchmark starts all producers and consumers coroutines. Producers send to a channel messages and then do
+ * some work on CPU, consumers receive messages from the channel and do some work on CPU. Total amount of sent messages
+ * is ~ [APPROXIMATE_BATCH_SIZE]. The main thread waits for all coroutines to stop, then measures the execution time of
+ * the current iteration.
+ *
+ * The benchmark stops execution the current benchmark configuration if execution times became stable and stopped changing
+ * drastically.
+ */
+fun main() {
```
Comment (ndkoval) [inline benchmarks/src/main/kotlin/benchmarks/channel/ChannelProducerConsumerMonteCarloBenchmark.kt]: It is better to create new JVM instances for each configuration

File: benchmarks/src/main/kotlin/macrobenchmarks/ChannelProducerConsumerMonteCarloBenchmark.kt [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+
+    val dispatcher = channelProducerConsumerBenchmarkWorker.dispatcher
+    if (dispatcher is Closeable) {
+        dispatcher.close()
+    }
+
+    return (endTime - startTime) / 1_000_000 // ms
+}
+
+class ChannelProducerConsumerBenchmarkIterationMonteCarlo(channelCreator: ChannelCreator,
```
Comment (qwwdfsad) [inline benchmarks/src/main/kotlin/macrobenchmarks/ChannelProducerConsumerMonteCarloBenchmark.kt]: It would be nice to have much shorter class names. It is really hard to follow the code when reading `ChannelProducerConsumerBenchmarkIterationMonteCarlo(...ten params...) : ChannelProducerConsumerBenchmarkIteration`

File: benchmarks/src/main/kotlin/macrobenchmarks/ChannelProducerConsumerMonteCarloBenchmark.kt [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+ */
+private val WITH_SELECT = listOf(true, false)
+/**
+ * Benchmark output file
+ */
+private const val OUTPUT = "out/results_channel_producer_consumer_montecarlo.csv"
+/**
+ * Options for the new jvm instance
+ */
+private val jvmOptions = listOf<String>(/*"-Xmx64m", "-XX:+PrintGC"*/)
```
Comment (qwwdfsad) [inline benchmarks/src/main/kotlin/macrobenchmarks/ChannelProducerConsumerMonteCarloBenchmark.kt]: Current default configuration is unacceptable. 
On 16 cores MBP the running time is estimated to be 9 hours. Nobody will ever run this benchmark is such a configuration. It should have a developer mode (from 1 to 10 minutes max running time) and "measurement mode" with 30-60 minutes max on a decent machine.

File: benchmarks/src/main/kotlin/macrobenchmarks/ChannelProducerConsumerMonteCarloBenchmark.kt [hunk 20]
```diff
@@ -0,0 +1,264 @@
...
+            val startTime = args[7].toLong()
+
+            print("\rchannel=$channel threads=$threads withBalancing=$withBalancing dispatcherType=$dispatcherType withSelect=$withSelect: warm-up phase... [${eta(currentConfigurationNumber, totalIterations, startTime)}]")
+
+            repeat(WARM_UP_ITERATIONS) {
+                runIteration(threads, channel, withBalancing, withSelect, dispatcherType)
+            }
+
+            runMonteCarlo(threads, channel, withBalancing, withSelect, dispatcherType) {
+                eta(currentConfigurationNumber, totalIterations, startTime)
```
Comment (qwwdfsad) [inline benchmarks/src/main/kotlin/macrobenchmarks/ChannelProducerConsumerMonteCarloBenchmark.kt]: Please also print benchmark score for each intermediate configuration (in addition to configuration and ETA itself)

Comment (ndkoval) [reply benchmarks/src/main/kotlin/macrobenchmarks/ChannelProducerConsumerMonteCarloBenchmark.kt]: we already print this information

File: benchmarks/src/main/kotlin/benchmarks/common/ChannelProdConsBenchmarkIteration.kt [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+        }
+        // Wait until the work is done
+        phaser.arriveAndAwaitAdvance()
+    }
+
+    private suspend fun produce(element: Int, dummy: Channel<Int>?, coroutineNumber: Int) {
+        if (withSelect) {
+            select<Unit> {
+                channel.onSend(element) {}
+                dummy!!.onReceive {}
```
Comment (qwwdfsad) [inline benchmarks/src/main/kotlin/benchmarks/common/ChannelProdConsBenchmarkIteration.kt:59]: Please documents what's the purpose of dummy

File: benchmarks/src/main/kotlin/benchmarks/common/ChannelProdConsBenchmarkIteration.kt [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+
+import kotlinx.coroutines.*
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.selects.*
+import java.util.concurrent.*
+
+/**
+ * Runs a batch send-receive pairs under the specified workload.
+ */
+public abstract class ChannelProdConsBenchmarkIteration(
```
Comment (qwwdfsad) [inline benchmarks/src/main/kotlin/benchmarks/common/ChannelProdConsBenchmarkIteration.kt:15]: Currently, the only possible way to understand the structure of the benchmark (what does it measure, how does it split the workload, what's the relationship between producers and consumers, what caveats does it have etc.) can be understood only from studying the source code, which doesn't seem really maintainable.

Please describe the top-level process in the root KDoc
